### PR TITLE
在smooth_l1_loss函数中加入is_huber参数

### PR DESCRIPTION
**Pull request by Qin-sx**
### PR Category

User Experience

### PR Types

Improvements 

### Description

paddle.nn.SmoothL1Loss和torch.nn.SmoothL1Loss功能不一致。  
paddle.nn.SmoothL1Loss的实现为huber loss，与torch.nn.HuberLoss对应。  
torch.nn.SmoothL1Loss的实现为huber loss再除以beta。  

在paddle.nn.SmoothL1Loss中加入is_huber参数:  
- is_huber=True (default)，对应torch.nn.HuberLoss
- is_huber=False，对应torch.nn.HuberLoss

[调研](https://github.com/Qin-sx/Paddle/blob/SmoothL1Loss/SmoothL1Loss.md)

### CODE REVIEW

**Comment by Qin-sx** [comment]
docs: https://github.com/PaddlePaddle/docs/pull/7323  
PaConvert: https://github.com/PaddlePaddle/PaConvert/pull/593

**Comment by zhwesky2010** [comment]
> * is_huber=False

is_huber=False，对应torch.nn.SmoothL1Loss

**Review by zhwesky2010 — COMMENTED** [review]
这个改动应该不需要这么大吧？无需新增kernel，复用huber_loss的kernel

is_huber=True时，不进行任何操作，is_huber=False时，在原kernel的基础上除以delta

**Comment by Qin-sx** [comment]
> 这个改动应该不需要这么大吧？无需新增kernel，复用huber_loss的kernel
> 
> is_huber=True时，不进行任何操作，is_huber=False时，在原kernel的基础上除以delta

嗯，因为我看pytorch是有两个kernel，感觉应该是底层kernel计算效率更高一些。
```cpp
void smooth_l1_kernel_cuda(TensorIteratorBase& iter, double beta) {
  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "smooth_l1_cuda", [&iter, beta]() {
    scalar_t beta_val(beta);
    gpu_kernel(iter, [beta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
      auto z = ::abs(a - b);
      return z < beta_val ? scalar_t(0.5) * z * z / beta_val : z - scalar_t(0.5) * beta_val;
    });
  });
}

void huber_kernel_cuda(TensorIterator& iter, double delta) {
  AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "huber_cuda", [&iter, delta] {
    scalar_t delta_val(delta);
    gpu_kernel(iter, [delta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
      auto z = ::abs(a - b);
      return z < delta_val ? scalar_t(0.5) * z * z : delta_val * (z - scalar_t(0.5) * delta_val);
    });
  });
}
```
想麻烦问一下，这个修改是直接在python函数中修改kernel返回的结果吗？

**Comment by zhwesky2010** [comment]
> > 这个改动应该不需要这么大吧？无需新增kernel，复用huber_loss的kernel
> > is_huber=True时，不进行任何操作，is_huber=False时，在原kernel的基础上除以delta
> 
> 嗯，因为我看pytorch是有两个kernel，感觉应该是底层kernel计算效率更高一些。
> 
> ```c++
> void smooth_l1_kernel_cuda(TensorIteratorBase& iter, double beta) {
>   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "smooth_l1_cuda", [&iter, beta]() {
>     scalar_t beta_val(beta);
>     gpu_kernel(iter, [beta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
>       auto z = ::abs(a - b);
>       return z < beta_val ? scalar_t(0.5) * z * z / beta_val : z - scalar_t(0.5) * beta_val;
>     });
>   });
> }
> 
> void huber_kernel_cuda(TensorIterator& iter, double delta) {
>   AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "huber_cuda", [&iter, delta] {
>     scalar_t delta_val(delta);
>     gpu_kernel(iter, [delta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
>       auto z = ::abs(a - b);
>       return z < delta_val ? scalar_t(0.5) * z * z : delta_val * (z - scalar_t(0.5) * delta_val);
>     });
>   });
> }
> ```
> 
> 想麻烦问一下，这个修改是直接在python函数中修改kernel返回的结果吗？

这个PR先留着，修改API：`nn.functionla.smooth_l1_loss`中的实现，代码最后面增加一个除法操作

infoflow 2025-06-17 12-15-57

**Comment by Qin-sx** [comment]
> > > 这个改动应该不需要这么大吧？无需新增kernel，复用huber_loss的kernel
> > > is_huber=True时，不进行任何操作，is_huber=False时，在原kernel的基础上除以delta
> > 
> > 
> > 嗯，因为我看pytorch是有两个kernel，感觉应该是底层kernel计算效率更高一些。
> > ```c++
> > void smooth_l1_kernel_cuda(TensorIteratorBase& iter, double beta) {
> >   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "smooth_l1_cuda", [&iter, beta]() {
> >     scalar_t beta_val(beta);
> >     gpu_kernel(iter, [beta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
> >       auto z = ::abs(a - b);
> >       return z < beta_val ? scalar_t(0.5) * z * z / beta_val : z - scalar_t(0.5) * beta_val;
> >     });
> >   });
> > }
> > 
> > void huber_kernel_cuda(TensorIterator& iter, double delta) {
> >   AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, iter.dtype(), "huber_cuda", [&iter, delta] {
> >     scalar_t delta_val(delta);
> >     gpu_kernel(iter, [delta_val] GPU_LAMBDA (scalar_t a, scalar_t b) -> scalar_t {
> >       auto z = ::abs(a - b);
> >       return z < delta_val ? scalar_t(0.5) * z * z : delta_val * (z - scalar_t(0.5) * delta_val);
> >     });
> >   });
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 想麻烦问一下，这个修改是直接在python函数中修改kernel返回的结果吗？
> 
> 这个PR先留着，修改API：`nn.functionla.smooth_l1_loss`中的实现，代码最后面增加一个除法操作
> 
> infoflow 2025-06-17 12-15-57

嗯，收到，已修改  
https://github.com/PaddlePaddle/Paddle/pull/73416
请问这个PR是否修改底层算子函数名为smooth_l1_loss后再提交？
